File: takengine/takengine/src/main/java/com/atakmap/spatial/QuadBlob.java
package com.atakmap.spatial;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import com.atakmap.coremap.maps.coords.GeoPoint;
import com.atakmap.math.MathUtils;

/**
 * Convenience class for generating SpatiaLite Geometry blobs for quadrilaterals
 * (geometry type POLYGON).
 * 
 * <P>This class is NOT thread-safe.
 * 
 * @author Developer
 */
public final class QuadBlob {
    private final static int QUAD_BLOB_SIZE = 132;

    private ByteBuffer quad;
    private byte blobEndian;

    public QuadBlob() {
        this(ByteOrder.nativeOrder());
    }

    public QuadBlob(ByteOrder endian) {
        this.quad = null;
        
        this.blobEndian = (endian == ByteOrder.BIG_ENDIAN) ? (byte)0x00 : (byte)0x01;
    }

    /**
     * Returns the SpatiaLite geometry equivalent to the quadrilateral specified
     * by the four points. The returned array is <I>live</I> and must be copied
     * if this method is expected to be called again before use of the data is
     * complete.
     */
    public byte[] getBlob(GeoPoint a, GeoPoint b, GeoPoint c, GeoPoint d) {
        if(this.quad == null) {
            this.quad = ByteBuffer.wrap(new byte[QUAD_BLOB_SIZE]);
            
            this.quad.order(ByteOrder.nativeOrder());
            
            this.quad.put((byte)0x00);
            this.quad.put(blobEndian);
            this.quad.putInt(4326);
            this.quad.position(this.quad.position()+32);
            this.quad.put((byte)0x7C);
            this.quad.putInt(0x03);
            
            this.quad.putInt(1);
            this.quad.putInt(5);
            
            this.quad.put(QUAD_BLOB_SIZE-1, (byte)0xFE);
        }

        final double alat = a.getLatitude();
        final double alng = a.getLongitude();
        final double blat = b.getLatitude();
        final double blng = b.getLongitude();
        final double clat = c.getLatitude();
        final double clng = c.getLongitude();
        final double dlat = d.getLatitude();
        final double dlng = d.getLongitude();
        
        this.quad.putDouble(51, alng);
        this.quad.putDouble(59, alat);
        this.quad.putDouble(67, blng);
        this.quad.putDouble(75, blat);
        this.quad.putDouble(83, clng);
        this.quad.putDouble(91, clat);
        this.quad.putDouble(99, dlng);
        this.quad.putDouble(107, dlat);
        this.quad.putDouble(115, alng);
        this.quad.putDouble(123, alat);

        final double mbrMinX = MathUtils.min(alng, blng, clng, dlng);
        final double mbrMinY = MathUtils.min(alat, blat, clat, dlat);
        final double mbrMaxX = MathUtils.max(alng, blng, clng, dlng);
        final double mbrMaxY = MathUtils.max(alat, blat, clat, dlat);
        
        this.quad.putDouble(6, mbrMinX);
        this.quad.putDouble(14, mbrMinY);
        this.quad.putDouble(22, mbrMaxX);
        this.quad.putDouble(30, mbrMaxY);
        
        return this.quad.array();
    }
}

File: takkernel/engine/src/main/cpp/feature/QuadBlob.cpp
#include "feature/QuadBlob.h"

#include "feature/Point.h"
#include "util/IO.h"

using namespace atakmap::feature;

using namespace atakmap::util;

#define QUAD_BLOB_SIZE 132

namespace
{
    void initBlob(uint8_t *quadblob, Endian endian);

    template<class T>
    struct ByteBuffer
    {
        uint8_t *(*put)(uint8_t *buf, size_t idx, T val);
    };

    template<class T>
    uint8_t *putNative(uint8_t *buf, size_t idx, T val);

    template<class T>
    uint8_t *putNativeReverse(uint8_t *buf, size_t idx, T val);

    template<class T>
    ByteBuffer<T> getByteBuffer(Endian endian);

    double min4(double a, double b, double c, double d);
    double max4(double a, double b, double c, double d);
}

QuadBlob::QuadBlob() :
    endian(PlatformEndian)
{
    initBlob(this->quad, endian);
}

QuadBlob::QuadBlob(Endian endian_) :
    endian(endian_)
{
    initBlob(this->quad, endian);
}

uint8_t *QuadBlob::getBlob(const Point *a, const Point *b, const Point *c, const Point *d)
{
    ByteBuffer<double> doubleBuffer = getByteBuffer<double>(this->endian);

    const double alat = a->y;
    const double alng = a->x;
    const double blat = b->y;
    const double blng = b->x;
    const double clat = c->y;
    const double clng = c->x;
    const double dlat = d->y;
    const double dlng = d->x;

    doubleBuffer.put(this->quad, 51, alng);
    doubleBuffer.put(this->quad, 59, alat);
    doubleBuffer.put(this->quad, 67, blng);
    doubleBuffer.put(this->quad, 75, blat);
    doubleBuffer.put(this->quad, 83, clng);
    doubleBuffer.put(this->quad, 91, clat);
    doubleBuffer.put(this->quad, 99, dlng);
    doubleBuffer.put(this->quad, 107, dlat);
    doubleBuffer.put(this->quad, 115, alng);
    doubleBuffer.put(this->quad, 123, alat);

    const double mbrMinX = min4(alng, blng, clng, dlng);
    const double mbrMinY = min4(alat, blat, clat, dlat);
    const double mbrMaxX = max4(alng, blng, clng, dlng);
    const double mbrMaxY = max4(alat, blat, clat, dlat);

    doubleBuffer.put(this->quad, 6, mbrMinX);
    doubleBuffer.put(this->quad, 14, mbrMinY);
    doubleBuffer.put(this->quad, 22, mbrMaxX);
    doubleBuffer.put(this->quad, 30, mbrMaxY);

    return this->quad;
}

size_t QuadBlob::getQuadBlobSize()
{
    return QUAD_BLOB_SIZE;
}

namespace
{
    void initBlob(uint8_t *quadblob, Endian endian)
    {
        ByteBuffer<int32_t> intBuffer = getByteBuffer<int32_t>(endian);

        const uint8_t blobEndian = (endian == Endian::BIG_ENDIAN) ? 0x00 : 0x01;

        uint8_t *pQuadblob = quadblob;

        *pQuadblob++ = (uint8_t)0x00;
        *pQuadblob++ = blobEndian;
        pQuadblob = intBuffer.put(pQuadblob, 0, 4326);
        pQuadblob += 32;
        *pQuadblob++ = (uint8_t)0x7C;
        pQuadblob = intBuffer.put(pQuadblob, 0, 0x03);

        pQuadblob = intBuffer.put(pQuadblob, 0, 1);
        pQuadblob = intBuffer.put(pQuadblob, 0, 5);

        quadblob[QUAD_BLOB_SIZE - 1] = 0xFE;
    }

    template<class T>
    uint8_t *putNative(uint8_t *buf, size_t idx, T val)
    {
        T *tBuf = reinterpret_cast<T *>(buf + idx);
        *tBuf = val;
        return buf + idx + sizeof(T);
    }

    template<class T>
    uint8_t *putNativeReverse(uint8_t *buf, size_t idx, T val)
    {
        const size_t retval = sizeof(T);
        const auto *valPtr = reinterpret_cast<const uint8_t *>(&val);
        for (size_t i = 0; i < retval; i++) {
            buf[i+idx] = valPtr[retval - i - 1];
        }
        return buf + idx + retval;
    }

    template<class T>
    ByteBuffer<T> getByteBuffer(Endian endian)
    {
        ByteBuffer<T> retval;
        retval.put = (PlatformEndian == endian) ? putNative<T> : putNativeReverse<T>;
        return retval;
    }

    double min4(double a, double b, double c, double d)
    {
        double retval = a;
        if (retval > b)
            retval = b;
        if (retval > c)
            retval = c;
        if (retval > d)
            retval = d;
        return retval;
    }

    double max4(double a, double b, double c, double d)
    {
        double retval = a;
        if (retval < b)
            retval = b;
        if (retval < c)
            retval = c;
        if (retval < d)
            retval = d;
        return retval;
    }
}

